src/sinks/datadog/logs.rs

```diff
@@ -66,31 +130,223 @@ impl SinkConfig for DatadogLogsConfig {
     }
 }
 
-fn encode_event(
-    mut event: Event,
-    api_key: &str,
-    encoding: &EncodingConfig<Encoding>,
-) -> Option<Bytes> {
-    let log = event.as_mut_log();
+#[async_trait::async_trait]
+impl HttpSink for DatadogLogsJsonService {
+    type Input = serde_json::Value;
+    type Output = Vec<BoxedRawValue>;
+
+    fn encode_event(&self, mut event: Event) -> Option<Self::Input> {
+        let log = event.as_mut_log();
+
+        if let Some(message) = log.remove(&log_schema().message_key()) {
+            log.insert("message", message);
+        }
+
+        if let Some(timestamp) = log.remove(&log_schema().timestamp_key()) {
+            log.insert("date", timestamp);
+        }
 
-    if let Some(message) = log.remove(&log_schema().message_key()) {
-        log.insert("message", message);
+        if let Some(host) = log.remove(&log_schema().host_key()) {
+            log.insert("host", host);
+        }
+
+        self.config.encoding.apply_rules(&mut event);
+
+        Some(json!(event.into_log()))
+    }
+
+    async fn build_request(&self, events: Self::Output) -> crate::Result<http::Request<Vec<u8>>> {
+        let body = serde_json::to_vec(&events)?;
+        build_request(&self.config, body)
     }
+}
+
+#[async_trait::async_trait]
+impl HttpSink for DatadogLogsTextService {
+    type Input = Bytes;
+    type Output = Vec<Bytes>;
 
-    if let Some(timestamp) = log.remove(&log_schema().timestamp_key()) {
-        log.insert("date", timestamp);
+    fn encode_event(&self, event: Event) -> Option<Self::Input> {
+        encode_event(event, &self.config.encoding)
     }
 
-    if let Some(host) = log.remove(&log_schema().host_key()) {
-        log.insert("host", host);
+    async fn build_request(&self, events: Self::Output) -> crate::Result<http::Request<Vec<u8>>> {
+        let body: Vec<u8> = events.iter().flat_map(|b| b.into_iter()).cloned().collect();
+
+        build_request(&self.config, body)
     }
+}
+
+/// Build the request, GZipping the contents if the config specifies.
+fn build_request(
+    config: &DatadogLogsConfig,
+    body: Vec<u8>,
+) -> crate::Result<http::Request<Vec<u8>>> {
```

Comment:
Since this takes `DatadogLogsConfig` as the first parameter, what do you think of moving this into a method of that struct?

Reply:
Seems logical.